File: src/graphics/cpp/image_cropper/image_cropper.cpp
#include "image_cropper.h"
#include <QDebug>
#include <QMessageBox>
#include <QMimeData>
#include <QPainter>
#include <QPainterPath>
#include <QWheelEvent>
#include <QtMath>

ImageCropper::ImageCropper(QWidget *parent) : QWidget(parent) {
  // accept droped images
  setAcceptDrops(true);
}

void ImageCropper::paintEvent(QPaintEvent *event) {
  Q_UNUSED(event);

  QPainter painter;

  // draw a black circle in the center of the widget
  // with empty fill
  painter.begin(this);
  painter.setPen(Qt::black);
  painter.setBrush(Qt::NoBrush);
  painter.drawEllipse(rect().center(), rect().width() / 2, rect().height() / 2);
  painter.end();

  if (pixmap.isNull()) {
    // draw text: "Drag and drop an image here"
    painter.begin(this);
    painter.setPen(Qt::black);
    painter.setBrush(Qt::NoBrush);
    painter.drawText(rect(), Qt::AlignCenter, "Drag and drop an image here");
    painter.end();
  }

  else {
    // draw pixmap within the circle drawn above
    // it should not go out of the circle bounds
    // the picture has to be cropped with clip path ellipse
    QPainterPath path;
    path.addEllipse(rect().center(), rect().width() / 2, rect().height() / 2);
    painter.begin(this);
    painter.setClipPath(path);
    painter.drawPixmap(rect(), pixmap);
    painter.end();
  }
}

void ImageCropper::dropEvent(QDropEvent *event) {
  // get the file name of the dropped image
  auto fileName = event->mimeData()->urls().first().toLocalFile();

  // set the pixmap from the file
  setPixmapFromFile(fileName);
}

void ImageCropper::setPixmapFromFile(QString fileName) {
  QPixmap newPix(fileName);

  if (newPix.isNull()) {
    QMessageBox::warning(this, tr("Invalid Image"),
                         tr("You tried loading an invalid file."));
    return;
  }

  pixmap = newPix;
  update();
}

void ImageCropper::dragEnterEvent(QDragEnterEvent *event) {
  // accept the event if the mime data contains a url
  if (event->mimeData()->hasUrls()) {
    event->acceptProposedAction();
  }
}

void ImageCropper::dragMoveEvent(QDragMoveEvent *event) {
  // accept the event if the mime data contains a url
  if (event->mimeData()->hasUrls()) {
    event->acceptProposedAction();
  }
}

File: src/graphics/python/image_cropper/image_cropper.py
from PyQt6 import QtWidgets, uic, QtGui
from PyQt6.QtCore import Qt, QRectF
from PyQt6.QtGui import QPixmap, QPen, QColor
from PyQt6.QtWidgets import QMessageBox


class ImageCropper(QtWidgets.QWidget):
    def __init__(self, parent=None):
        super(ImageCropper, self).__init__(parent)
        self.setAcceptDrops(True)
        self.pixmap = QPixmap()

    def paintEvent(self, event):
        painter = QtGui.QPainter()
        painter.begin(self)
        painter.setPen(QPen(QColor("black")))
        painter.setBrush(Qt.BrushStyle.NoBrush)
        painter.drawEllipse(self.rect().center(), self.rect().width() / 2, self.rect().height() / 2)
        painter.end()
        if self.pixmap.isNull():
            painter.begin(self)
            painter.setPen(QPen(QColor("black")))
            painter.setBrush(Qt.BrushStyle.NoBrush)
            painter.drawText(self.rect(), Qt.AlignmentFlag.AlignCenter, "Drag and drop an image here")
            painter.end()
        else:
            painter.begin(self)
            path = QtGui.QPainterPath()
            path.addEllipse(QRectF(self.rect()))
            painter.setClipPath(path)
            painter.drawPixmap(self.rect(), self.pixmap)
            painter.end()
    
    def dropEvent(self, event):
        urls = event.mimeData().urls()
        if len(urls) == 0:
            return
        file_name = urls[0].toLocalFile()
        self.set_pixmap_from_file(file_name)
    
    def set_pixmap_from_file(self, file_name):
        new_pix = QtGui.QPixmap(file_name)
        if new_pix.isNull():
            QMessageBox.warning(self, "Invalid Image",
                                 "You tried loading an invalid file.")
            return
        self.pixmap = new_pix
        self.update()

    def dragEnterEvent(self, event):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
    
    def dragMoveEvent(self, event):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
    